blog/2024-07-3-uppy-4.0.md

```diff
@@ -0,0 +1,364 @@
+---
+title:
+  'New Uppy 4.0 major: TypeScript rewrite, Google Photos, React hooks, and much
+  more.'
+date: 2024-07-03
+authors: [aduh95, evgenia, mifi, murderlon]
+image: 'https://uppy.io/img/blog/3.13-3.21/dog-coding-laptop-mars-christmas-tree-2.jpg'
+slug: 'uppy-4.0'
+published: false
+toc_max_heading_level: 2
+---
+
+Hi ha, this is some goofy introduction.
+
+## TypeScript rewrite
+
+In the year 2024 people expect excellent types from their libraries. We used to
+author types separately by hand but they were often inconsistent or incomplete.
+Now Uppy has been completely rewritten in TypeScript!
+
+From now on you’ll be in safe hands when working with Uppy, whether it’s setting
+the right options, building plugins, or listening to events.
+
+```ts
+import Uppy from '@uppy/core';
+
+const uppy = new Uppy();
+
+// Event name autocompletion and inferred argument types
+uppy.on('file-added', (file) => {
+	console.log(file);
+});
+```
+
+One important thing to note are the new generics on `@uppy/core`.
+
+```ts
+import Uppy from '@uppy/core';
+// xhr-upload is for uploading to your own backend.
+import XHRUpload from '@uppy/xhr-upload';
+
+// Your own metadata on files
+type Meta = { myCustomMetadata: string };
+// The response from your server
+type Body = { someThingMyBackendReturns: string };
+
+const uppy = new Uppy<Meta, Body>().use(XHRUpload, {
+	endpoint: '/upload',
+});
+
+const id = uppy.addFile({
+	name: 'example.jpg',
+	data: new Blob(),
+	meta: { myCustomMetadata: 'foo' },
+});
+
+// This is now typed
+const { myCustomMetadata } = uppy.getFile(id).meta;
+
+await uppy.upload();
+
+// This is strictly typed too
+const { someThingMyBackendReturns } = uppy.getFile(id).response;
```

Comment:
Getting the following TS errors:

```
index.ts:3:23 - error TS2307: Cannot find module '@uppy/xhr-upload' or its corresponding type declarations.

3 import XHRUpload from '@uppy/xhr-upload';
                        ~~~~~~~~~~~~~~~~~~

index.ts:26:9 - error TS2339: Property 'someThingMyBackendReturns' does not exist on type '{ body?: Body; status: number; bytesUploaded?: number; uploadURL?: string; }'.

26 const { someThingMyBackendReturns } = uppy.getFile(id).response;
           ~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in the same file, starting at: index.ts:3
```

Reply:
Where? CI is green.

Reply:
Fix is https://github.com/transloadit/uppy.io/pull/258/files#r1670167026